core/checks.go

```diff
@@ -253,3 +195,210 @@ func (c *Check) Query() []dagql.Selector {
+
+	// We can't distinguish legitimate errors from failed checks, so we just discard.
+	// Bubbling them up to here makes telemetry more useful (no green when a check failed)
+	return c, nil
+}
+
+func (c *Check) run(
+	ctx context.Context,
+	dag *dagql.Server,
+	tryScaleOut bool,
+) (rerr error) {
+	if tryScaleOut {
+		if ok, err := c.tryScaleOut(ctx); ok {
+			return err
+		}
+	}
+
+	selectPath := []dagql.Selector{{Field: gqlFieldName(c.Module.Name())}}
+	for _, field := range c.Path {
+		selectPath = append(selectPath, dagql.Selector{Field: field})
+	}
+
+	var checkParent dagql.AnyObjectResult
+	if err := (func(ctx context.Context) (rerr error) {
+		ctx, span := Tracer(ctx).Start(ctx, "load check context",
+			// Prevent logs from bubbling up past this point.
+			telemetry.Boundary(),
+			// We're only using this span as a log encapsulation boundary; show
+			// its child spans inline.
+			telemetry.Passthrough(),
+		)
+		defer telemetry.EndWithCause(span, &rerr)
+		return dag.Select(ctx, dag.Root(), &checkParent, selectPath[:len(selectPath)-1]...)
+	})(ctx); err != nil {
+		return err
+	}
+
+	// Call the check
+	var status dagql.AnyResult
+	if err := dag.Select(dagql.WithRepeatedTelemetry(ctx), checkParent, &status, selectPath[len(selectPath)-1]); err != nil {
+		return err
+	}
+
+	if obj, ok := dagql.UnwrapAs[dagql.AnyObjectResult](status); ok {
+		// If the check returns a syncable type, sync it
+		if syncField, has := obj.ObjectType().FieldSpec("sync", dag.View); has {
+			if !syncField.Args.HasRequired(dag.View) {
+				if err := dag.Select(
+					dagql.WithRepeatedTelemetry(ctx),
+					obj,
+					&status,
+					dagql.Selector{Field: "sync"},
+				); err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func (c *Check) tryScaleOut(ctx context.Context) (_ bool, rerr error) {
+	q, err := CurrentQuery(ctx)
+	if err != nil {
+		return true, err
+	}
+
+	cloudEngineClient, useCloudEngine, err := q.CloudEngineClient(ctx,
+		c.Module.Source.Value.Self().AsString(),
+		// FIXME: we're saying the "function" is the check and no execCmd,
+		// which works with cloud but is weird
+		c.Name(),
+		nil,
+	)
+	if err != nil {
+		return true, fmt.Errorf("engine-to-engine connect: %w", err)
+	}
+	if !useCloudEngine {
+		return false, nil
+	}
+	defer func() {
+		rerr = errors.Join(rerr, cloudEngineClient.Close())
+	}()
+
+	query := cloudEngineClient.Dagger().QueryBuilder()
+
+	switch c.Module.Source.Value.Self().Kind {
+	case ModuleSourceKindLocal:
+		query = query.Select("moduleSource").
+			Arg("refString", filepath.Join(
+				c.Module.Source.Value.Self().Local.ContextDirectoryPath,
+				c.Module.Source.Value.Self().SourceRootSubpath,
+			))
+	case ModuleSourceKindGit:
+		query = query.Select("moduleSource").
+			Arg("refString", c.Module.Source.Value.Self().AsString()).
+			Arg("refPin", c.Module.Source.Value.Self().Git.Commit).
+			Arg("requireKind", c.Module.Source.Value.Self().Kind)
+	case ModuleSourceKindDir:
```

Comment:
In which case this may happen? If I remember well the `ModuleSourceKindDir` is when we do `Directory.AsModuleSource` so it should never lead to an actual execution no?